### ISSUE TITLE
Several incorrect parsing errors

### ISSUE THREAD

**Issue by bjorn3** [rust-analyzer/rust-analyzer#919]
```rust
// ===== fixed =====
extern "C" {
        pub fn printf(format: *const i8, ...) -> i32;
//                                      ^^  ^ ^     ^     8 different errors at the specified locations
}
```

```rust
pub fn abc() {
    #[cfg(debug_assertions)]
//                          ^ expected a let statement or an item after attributes in block
    self.some_method();
}
```

```rust
pub fn bcd() {
    let place = if is_ssa {
        #[cfg(debug_assertions)]
//      ^^^^^^^^^^^^^^^^^^^^^^^^ A block in this position cannot accept inner attributes
//                              ^ expected a let statement or an item after attributes in block
        {
            abc();
        }
    } else {};
}
```

The following ones only works on nightly:

```rust
pub macro unimpl($($tt:tt)*) {
//       ^ expected EXCL
//                          ^^ expected an item
    panic!(NonFatal(format!($($tt)*)));
}
```

```rust
fn def() {
    box 42;
//     ^ expected SEMI
}

**Comment by DJMcNab** [comment]
We explicitly do not handle macros 2.0 until they are closer to stabilisation I believe. I have taken over attribute parsing, but am unavailable to work on it for the next week.

However, that `extern "C"` example is interesting and needs investigation. I don't think we handle the `box` contextual keyword, and I imagine it needs care (especially if `box` is a legal identifier)

**Comment by bjorn3** [comment]
> We explicitly do not handle macros 2.0 until they are closer to stabilisation I believe.

Got it.

> I don't think we handle the box contextual keyword, and I imagine it needs care (especially if box is a legal identifier)

It is a keyword. It may be removed in the future though, just wanted to mention it.

**Comment by DJMcNab** [comment]
Cool - I suppose box will be fairly simple then, although handling it in `ty` will require lang item collection, which I'm not sure is implemented fully yet.

**Comment by bjorn3** [comment]
Found some more. These are from rustc:

```rust
use std::fmt::Write as _;
//                     ^ expected a name
```

```rust
    let _: io::Result<()> = try {
        let mut file = create_dump_file(tcx, "mir", pass_num, pass_name, disambiguator, source)?;
//      ^  ^^  ^    ^^^                ^^    ^                                                ^^^
    }
```

**Comment by matklad** [comment]
The vararg is fixed in #926 `macro` and `try` are unstable, but it would still be good to support them.

**Comment by matklad** [comment]
and #927 fixes `as _`.

**Comment by bjorn3** [comment]
Thanks!

**Comment by bjorn3** [comment]
Another one:

```rust
trait Abc {
    fn abc(mut b: u8) {}
//         ^ expected type
//            ^ expected COMMA
}
```

**Comment by kiljacken** [comment]
All the above examples that parse in rustc parse in rust-analyzer. Time to close this one?
